File: erdos_renyi/erdos.py
import random
import networkx as nx
import matplotlib.pyplot as plt

def erdos_renyi(nodes, edges):
    print nodes, edges
    adj_mat = [[0 for i in range(nodes)] for j in range(nodes)]
    p_link_formation = (2.0 * edges) / (nodes * (nodes-1))
    #print p_link_formation
    edge = 0

    G = nx.Graph()
    G.add_nodes_from([i for i in range(nodes)])

    for i in range(nodes):
        for j in range(i+1, nodes):
            r = random.random()
            #print r
            if r <= p_link_formation:
                adj_mat[i][j] = 1
                adj_mat[j][i] = 1
                G.add_edge(i, j)
                edge += 1

    return G, edge

def erdos_renyi2(nodes, edges):
    adj_mat = [[0 for i in range(nodes)] for j in range(nodes)]
    G = nx.Graph()
    G.add_nodes_from([i for i in range(nodes)])

    for i in range(edges):
        while(True):
            n1 = random.randint(0, nodes-1)
            n2 = random.randint(0, nodes-1)
            if n1 != n2 and not adj_mat[n1][n2]:
                adj_mat[n1][n2] = 1
                adj_mat[n2][n1] = 1
                G.add_edge(n1, n2)
                break
    return G

nodes = input("Enter number of nodes: ")
edges = input("Enter number of edges: ")
#G = erdos_renyi2(nodes, edges)
G, edge = erdos_renyi(nodes, edges)
print edge
nx.draw(G)
#plt.show()
plt.savefig("graph.png")

File: erdos_renyi/erdos.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
using namespace std;

int erdos_renyi(int nodes, int edges, vector<vector<bool> > &adj_mat)
{
    adj_mat.resize(nodes, vector<bool>(nodes, 0));
    double p_link_formation = (double) edges * 2 / ((double)nodes * (nodes-1));
    double r;
    srand(time(NULL));
    int edge = 0;

    for (int i=0; i < nodes; i++)
        for (int j=i+1; j < nodes; j++)
        {
            r = ((double) rand() / (RAND_MAX)); //0 < r < 1
            if (r <= p_link_formation)
            {
                adj_mat[i][j] = 1;
                adj_mat[j][i] = 1;
                edge++;
            }

        }
    return edge;
}

void erdos_renyi2(int nodes, int edges, vector<vector<bool> > &adj_mat)
{
    adj_mat.resize(nodes, vector<bool>(nodes, 0));
    int n1, n2;
    srand(time(NULL));
    int edge = 0;

    for (int i=0; i < edges; i++)
    {
        while(true)
        {
            n1 = rand()%nodes;
            n2 = rand()%nodes;
            if (n1 != n2 && !adj_mat[n1][n2])
            {
                adj_mat[n1][n2] = 1;
                adj_mat[n2][n1] = 1;
                edge++;
                break;
            }
        }
    }
}

int main()
{
    int nodes, edges, edge;
    vector<vector<bool> > graph;
    cout<<"Network Parameters: nodes and edges?: ";
    cin>>nodes>>edges;

    double avg=0;
    int iter = 100000;
    for(int k=0; k < iter; k++)
    {
        graph.clear();
        edge = erdos_renyi(nodes, edges, graph);
        //cout<<"edge: "<<edge<<'\n';
        avg += edge;
    }
    avg/=iter;
    cout<<"avg: "<<avg<<'\n';

    /*graph.clear();
    erdos_renyi(nodes, edges, graph);
    for (int i=0; i < nodes; i++)
    {
        for (int j=0; j < nodes; j++)
            cout<<graph[i][j]<<" ";
        cout<<'\n';
    }
    cout<<"edges: "<<edge<<'\n';*/

    return 0;
}